syntax = "proto3";

package proto.bank;

service Bank {
  // GetBalance returns the balance of a wallet.
  rpc GetBalance (GetBalanceRequest) returns (GetBalanceResponse) {}
  // Pay transfers some amount of assets from client wallet to destination
  // wallet.
  rpc Pay (PayRequest) returns (PayResponse) {}
  // Transfer transfers some amount of assets from client wallet to destination
  // wallet.
  rpc Transfer (TransferRequest) returns (TransferResponse) {}
  // KillServer tells the server to suicide. This exists to facilitate testing,
  // and because this project is just a proof of concept.
  rpc KillServer (KillServerRequest) returns (KillServerResponse) {}
}

message GetBalanceRequest {
  string walletId = 1;
}

message GetBalanceResponse {
  int32 balance = 1;
}

message PayRequest {
  string walletId = 1;
  uint32 amount = 2;
}

message PayResponse {
  bytes orderId = 1;
  int32 status = 2;
}

message TransferRequest {
  string targetWalletId = 1;
  bytes orderId = 2;
  uint32 amount = 3;
}

message TransferResponse {
  int32 status = 1;
}

message KillServerRequest {}

message KillServerResponse {
  uint32 numAccounts = 1;
}
